File: dotnet/MiniThatchTest/InvoiceItemGeneratorTest.cs
namespace MiniThatchTest;
using MiniThatch;

public class InvoiceItemGeneratorTest : UnitTest
{
    protected override List<Action> Tests => new()
    {
        TestGenerateForTransactions,
    };

    private void TestGenerateForTransactions()
    {
        var account = new Account("Craig Ellis", 100.0);
        var txn1 = new Transaction("Insurance premium", "Insurance", 50);
        var txn2 = new Transaction("Dentist appointment", "copay", 80);
        var txn3 = new Transaction("Target purchase", "retail", 5);

        var generated = InvoiceItemGenerator.GenerateForTransactions(account, new List<Transaction> { txn1, txn2, txn3 });

        AssertEqual(generated[0].SpendType, "reimbursement");
        AssertEqual(generated[0].Amount, 50);
        AssertEqual(generated[0].Transaction.Name, "Insurance premium");

        AssertEqual(generated[1].SpendType, "reimbursement");
        AssertEqual(generated[1].Amount, 50);
        AssertEqual(generated[1].Transaction.Name, "Dentist appointment");

        AssertEqual(generated[2].SpendType, "overage");
        AssertEqual(generated[2].Amount, 30);
        AssertEqual(generated[2].Transaction.Name, "Dentist appointment");

        AssertEqual(generated[3].SpendType, "overage");
        AssertEqual(generated[3].Amount, 5);
        AssertEqual(generated[3].Transaction.Name, "Target purchase");
    }
}

File: python/invoice_item_generator_test.py
from account import Account
from invoice_item import InvoiceItem
from transaction import Transaction
from invoice_item_generator import InvoiceItemGenerator

def test_generate_for_transactions1():
  account = Account("Craig Ellis", 100)
  txn_1 = Transaction(name="Insurance premium", category="Insurance", amount=50)
  txn_2 = Transaction(name="Dentist appointment", category="copay", amount=80)
  txn_3 = Transaction(name="Target purchase", category="retail", amount=5)

  generated = InvoiceItemGenerator.generate_for_transactions(account, [txn_1, txn_2, txn_3])
  
  assert generated[0].spend_type == "reimbursement"
  assert generated[0].amount == 50
  assert generated[0].transaction.name == "Insurance premium"

  assert generated[1].spend_type == "reimbursement"
  assert generated[1].amount == 50
  assert generated[1].transaction.name == "Dentist appointment"

  assert generated[2].spend_type == "overage"
  assert generated[2].amount == 30
  assert generated[2].transaction.name == "Dentist appointment"

  assert generated[3].spend_type == "overage"
  assert generated[3].amount == 5
  assert generated[3].transaction.name == "Target purchase"


if __name__ == "__main__":
  test_generate_for_transactions1()

  print("All tests passed!")
